### ISSUE TITLE
How to increase the upload speed?

### ISSUE THREAD

**Issue by seabrus** [jalik/jalik-ufs#66]
Hello,

First of all, thanks a lot for your efforts and great package. It helped me to solve several problems with the previous file package which is now deprecated, unfortunately.

My question is How can I increase the upload speed?

I found those ways to set up the store options in the manual and example:
```javascript
    // (1) Default options -- at https://github.com/jalik/jalik-ufs/blob/master/ufs-uploader.js
    adaptive: true,
    capacity: 0.9,
    chunkSize: 16 * 1024,          // 16 Kb
    maxChunkSize: 4 * 1024 * 1000, // 4 Mb
    maxTries: 5,

    // (2) Options from the basic example in docs -- at https://github.com/jalik/jalik-ufs
    adaptive: true,
    capacity: 0.8,
    chunkSize: 8 * 1024,      // 8 Kb
    maxChunkSize: 128 * 1024, // 128 Kb
    maxTries: 5,

    // (3) Options from UFS-Example -- at https://github.com/jalik/ufs-example/blob/master/imports/client/templates/templates.js
      // const ONE_MB = 1024 * 100;
    adaptive: false,
    chunkSize: ONE_MB * 16.66, // 1.6 Mb
    maxChunkSize: ONE_MB * 20, // 2 Mb
    maxTries: 3
```
If I try to upload a file of 16Mb  in size. It takes me 15 seconds to upload it locally using the option set (1) and about 70 chunks like this one
```
POST http://localhost:3000/ufs/audios/wKeg6j3uDoZNmy8Yz -- [HTTP/1.1 204 No Content 81ms]
```
sent from the client. If I use the option set (3), it takes 7 seconds and 10 chunks. Keeping in mind that I plan to upload audio files up to 50Mb, the latter is quite acceptable, but 15 seconds looks a bit too many. 

However, the option set (3) is not adaptive (`adaptive: false`). In the manual you wrote that the adaptive approach is more effective and more preferrable.

Can you comment this situation? Are there ways to increase the upload speed? Maybe some other options, settings, or permissions should be taken into account? Should I prefer the adaptive approach anyway?
Any other recommendations will be appreciated, too.

Have a nice day :)

**Comment by jalik** [comment]
Hi @seabrus Thank you, I need to add details about that in the doc...

When `adaptive` is set to `true`, it means that the chunk size will automatically decrease if it takes too long to upload a chunk (on low bandwidth for example), but if it is fast enough it will increase until `maxChunkSize` is reached. So in this case `chunkSize` is the first size you want to start uploading with. It' a variable bitrate. This is the recommended way since it is kind of smart enough to optimize uploading speed.

When `adaptive` is set to `false`, it means that chunks size will never change, so if you set 1MB it will upload at 1MB/s whatever if your bandwidth can support this speed, it's a constant bitrate. This is good if you are on an intranet where you know every clients have the same bandwidth and so you can set the upload speed to be constant.

You can set the value you want but if the value is too high it can(will) freeze the browser.

By the way your issue made me find bugs that I've fixed in the `v0.6.6`, also I fixed the value of ONE_MB in the ufs-example app it was missing a `0` (`100` instead of `1000`) so maybe it's because of that your file was uploading slowly... I did test on my own and I can upload on localhost a **50MB+** file in less than **4 sec** with the following settings :

```js
let uploader = new UploadFS.Uploader({
                adaptive: true,
                chunkSize: ONE_MB,
                maxChunkSize: ONE_MB * 10,
                data: file,
                file: file,
                store: FileStore,
                maxTries: 3
            });
```

**Comment by seabrus** [comment]
Hi @jalik,

Thank you for the prompt answer. I'll try the new version and your recommendation a bit later today. Now I just want to specify my situation (I had to do that yesterday, sorry).

1). I use `jalik:jalik-ufs-gridfs` and here is my store:
```javascript
Audios = new Mongo.Collection('audios');

AudiosStore = new UploadFS.store.GridFS({
  collection: Audios,
  name: 'audios',
  chunkSize: 1024 * 255,

  filter: new UploadFS.Filter({
    minSize: 1,
    maxSize: 1024 * 1024 * 50, // 50 Mb
    contentTypes: ['audio/*', 'video/*'],
    extensions: ['oga', 'mp3', 'ogg'],
  }),

  permissions: new UploadFS.StorePermissions({
    insert: function (userId, doc) {
      return (userId && userId === doc.userId);
    },
    update: function (userId, doc) {
      return (userId && userId === doc.userId);
    },
    remove: function (userId, doc) {
      return (userId && userId === doc.userId);
    },
  }),

  onRead: function (fileId, file, req, res) {
    if (file.token !== req.query.token) {
      res.writeHead(403, { 'Content-Type': 'text/plain' });
      res.end('Forbidden');
      return false;
    }
    return true;
  },

  onFinishUpload: function (file) {
    console.log('AudiosStore: ' + file.name + ' has been uploaded');
  },
  onReadError: function (err, fileId, file) {
    console.error('AudiosStore: Cannot read ' + file.name);
  },
  onWriteError: function (err, fileId, file) {
    console.error('AudiosStore: Cannot write ' + file.name);
  },

});
```
2). Also, I explicitely declare
```javscript
Audios.deny({
  insert: function() { return true; },
  update: function() { return true; },
  remove: function() { return true; },
});
```
or
```javscript
Audios.deny({
  //insert: function() { return true; },
  //update: function() { return true; },
  remove: function() { return true; },
});
```
 --- here I follow the recommendations from the Meteor Guide to explicitely deny the client-side operations. As far as I could understand, now in your package all operations with DB are performed only from the server side, so I considered these declarations as safe. 
Also please note that the authors of CollectionFS emphasize in their documentation that all operations in CFS should be made from the client. So, for any one who has some experience in working with CFS, the question "only client-side?" or "only server-side?" or "both?" arises itself and needs some explanation. My experiments indicated that for `jalik-ufs 0.6.5` the answer is "server-side".
However, in both `Audios.deny()`'s cases  the result was almost the same, I didn't see difference in speed. Now I still use the latter version where only `remove` is denied explicitely. 

3). Moreover, I use `React` + `flow-router-ssr` + `MaterializeCSS`, so I select files with my own buttons and then make the following for the `selectedFiles` (or BLOBs) array:
```javscript
    for (let k = 0; k < selectedFiles.length; k++) {
      uploadUFSFile(selectedFiles[k], ...);
    }
```
where the function `uploadUFSFile` reproduces your function for uploadfing a file from the  `UploadFS.selectFiles()`:
```javscript
export const uploadUFSFile = function (newFile, ...) {
  let fileObj = {
    name: newFile.name,
    size: newFile.size,
    type: newFile.type,
    userId: Meteor.userId(),
    ... [some other properties go here]
  };

  let uploader = new UploadFS.Uploader({
    store: AudiosStore,
    data: newFile,
    file: fileObj,

    // The latest option set that I used
    adaptive: false,
    chunkSize: 1024 * 100 * 16.66, // 1.6 Mb
    maxChunkSize: 1024 * 100 * 20, // 2 Mb
    maxTries: 5,

    onError: function (err) {
      console.error(err);
    },
    onComplete: function (file) {
      console.log(file.name + ' has been uploaded');
    },
  });

  uploader.start();
};
```
I'm writing about all these details just in case that some of them could be important in speed tuning.
And yes, It could be great to see a special section in the documentation about uploading speed optimization and whether the operations with DB are server-side or client-side. 
Hope my case would be helpful, too :).
